### PR TITLE
Adding OpenSSL support

### PR DESCRIPTION
**Pull request by SebaLukas**
## Describe your changes
See title

## Issue ticket number and link
#19 

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: src/iso15118/tbd_controller.cpp [hunk 1]
```diff
@@ -87,7 +91,11 @@ void TbdController::handle_sdp_server_input() {
 
     auto connection = [this](bool secure_connection) -> std::unique_ptr<io::IConnection> {
         if (secure_connection) {
+#ifdef LIBISO_OPENSSL
             return std::make_unique<io::ConnectionSSL>(poll_manager, config.interface_name, config.ssl);
```
Comment (james-ctc) [inline src/iso15118/tbd_controller.cpp:95]: this #ifdef could be removed if a using statement was used after lines 12 and 14
e.g. "using ConnectionType = io::ConnectionSSL;" and "using ConnectionType = io::ConnectionTLS;"
return std::make_unique<ConnectionType>(poll_manager, config.interface_name, config.ssl);

**Review by james-ctc — APPROVED** [review]
Good separation between OpenSSL and MbedTLS.
I only have some minor suggestions.

File: src/iso15118/io/connection_ssl.cpp [hunk 2]
```diff
@@ -1,288 +1,373 @@
...
+
+Ipv6EndPoint ConnectionSSL::get_public_endpoint() const {
+    return end_point;
+}
+
+void ConnectionSSL::init_ssl(const config::SSLConfig& ssl_config) {
+
+    // Note: openssl does not provide support for ECDH-ECDSA-AES128-SHA256 anymore
+    static constexpr auto TLS1_2_CIPHERSUITES = "ECDHE-ECDSA-AES128-SHA256";
+    static constexpr auto TLS1_3_CIPHERSUITES = "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256";
```
[the remark this answers is not in the document]
Comment (SebaLukas) [reply src/iso15118/io/connection_ssl.cpp:165]: Good point, thx for noticing! I'll change that

File: src/iso15118/tbd_controller.cpp (hunk 1 above)
Comment (SebaLukas) [reply src/iso15118/tbd_controller.cpp:95]: Good suggestion! I'll change it like this

File: include/iso15118/io/connection_ssl.hpp [hunk 3]
```diff
@@ -14,11 +14,13 @@ namespace iso15118::io {
 struct SSLContext;
 class ConnectionSSL : public IConnection {
 public:
-    ConnectionSSL(PollManager&, const std::string& interface_name, const config::SSLConfig&);
+    ConnectionSSL(PollManager&, const std::string&, const config::SSLConfig&);
```
Comment (MarzellT) [inline include/iso15118/io/connection_ssl.hpp:17]: Why did you remove the parameter name here?

File: src/iso15118/io/connection_ssl.cpp [hunk 4]
```diff
@@ -1,288 +1,370 @@
...
-        [](void* callback_context, int debug_level, const char* file_name, int line_number, const char* message) {
-            logf("mbedtls debug (level: %d) - %s\n", debug_level, message);
-        },
-        stdout);
+ConnectionSSL::~ConnectionSSL() = default;
 
-    if (ssl_config.enable_ssl_logging) {
-        mbedtls_debug_set_threshold(3);
+void ConnectionSSL::set_event_callback(const ConnectionEventCallback& callback) {
+    this->event_callback = callback;
```
Comment (MarzellT) [inline src/iso15118/io/connection_ssl.cpp:154]: I feel like we should remove the `this->` part here for consistency

File: src/iso15118/io/connection_ssl.cpp [hunk 5]
```diff
@@ -1,288 +1,370 @@
...
+                    sendto(key_log_fd, line, strlen(line), 0, reinterpret_cast<const sockaddr*>(&destination_address),
+                           sizeof(destination_address));
+
+                if (udp_send_result != strlen(line)) {
+                    const auto error_msg = adding_err_msg("UDP send() failed");
+                    logf(error_msg.c_str());
+                }
+            }
+        });
+    }
```
Comment (MarzellT) [inline src/iso15118/io/connection_ssl.cpp:229]: Perhaps this could be a function call for better readability

**Review by MarzellT — APPROVED** [review]
I think this merge request looks really good

File: include/iso15118/io/connection_ssl.hpp (hunk 3 above)
Comment (SebaLukas) [reply include/iso15118/io/connection_ssl.hpp:17]: Since the other parameters also have no name.

File: src/iso15118/io/connection_ssl.cpp (hunk 4 above)
Comment (SebaLukas) [reply src/iso15118/io/connection_ssl.cpp:154]: Good point!

File: include/iso15118/io/connection_ssl.hpp (hunk 3 above)
Comment (MarzellT) [reply include/iso15118/io/connection_ssl.hpp:17]: That makes sense. I got a little bit confused here

File: src/iso15118/io/connection_ssl.cpp (hunk 5 above)
Comment (SebaLukas) [reply src/iso15118/io/connection_ssl.cpp]: With the `create_udp_socket` refactor, I don't think this needs to be in a separate function.

Comment (MarzellT) [reply src/iso15118/io/connection_ssl.cpp]: Alright!
